client/allocrunner/taskrunner/template/template.go

```diff
@@ -308,22 +312,22 @@ WAIT:
 				continue
 			}
 
-			dirty := false
 			for _, event := range events {
 				// This template hasn't been rendered
 				if event.LastWouldRender.IsZero() {
 					continue WAIT
 				}
-				if event.WouldRender && event.DidRender {
-					dirty = true
+				// If the template _actually_ rendered to disk, mark it dirty
+				if !event.LastDidRender.IsZero() {
+					dirtyEvents[event.Template.ID()] = event
 				}
```

Comment:
If I'm following this logic correctly, it looks like we can expect multiple events on after we get woken up by `TemplateRenderedCh` but we only want to "keep" any events that are actually rendered-to-disk. That way we're not going back to the `RenderedEvents` and in the `onTemplateRendered` handler and getting new events we haven't previously seen. 

If so, that makes sense to me but we could probably use a comment to that effect here as this area of code has clearly been a bug farm historically. Maybe something like:

```suggestion
				// If the template _actually_ rendered to disk, mark it
				// dirty. We track events here so that onTemplateRendered
				// doesn't go back to the runner's RenderedEvents and process
				// new events that don't make us dirty.				
				if !event.LastDidRender.IsZero() {
					dirtyEvents[event.Template.ID()] = event
				}
```

Reply:
While attempting to replicate this bug I able to "mostly" replicate the CT behavior from the GH issue fairly consistently (the _actual_ replication was a lot less consistent).  By "mostly" I mean on client restart, the task shouldn't have been restarting, because the event was being overwritten.. But it was restarting!  Turns out that the other template's event was triggering the dirty flag (it had `change_mode = "noop"`), and then in `onTemplateRendered` we were executing the `change_mode` of a template that was not actually dirty, and restarting.

Reply:
So technically `onTemplateRendered` was not getting _new_ events it hadn't seen, it was performing _different_ logic to determine if an event should trigger a change_mode than `handleFirstRender` was.  Probably the best fix here is to move all this logic into one location, but that's slightly bigger change.